src/pubsub/ua_pubsub_manager.c

```diff
@@ -707,20 +707,55 @@ UA_PubSubManager_generateUniqueGuid(UA_Server *server) {
     }
 }
 
-/* Initialization the PubSub configuration. */
-void
-UA_PubSubManager_init(UA_Server *server, UA_PubSubManager *pubSubManager) {
-    /* ToDo: Must be correctly generated from Mac address and port. */
+static UA_UInt64
+generateRandomUInt64(UA_Server *server) {
+    UA_UInt64 id = 0;
     UA_Guid *ident = NULL;
     ident = UA_Guid_new();
     if(!ident) {
         UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
                      "PubSub generate defaultPublisherId failed.");
-        return;
+        return 0;
     }
     *ident = UA_Guid_random();
-    pubSubManager->defaultPublisherId = *(UA_UInt64*)ident;
+    id = id + ident->data1;
+    id = (id << 32) + ident->data2;
+    id = (id << 16) + ident->data3;
     UA_free(ident);
+    return id;
+}
+
+/* Initialization the PubSub configuration. */
+void
+UA_PubSubManager_init(UA_Server *server, UA_PubSubManager *pubSubManager) {
+#ifdef UA_ARCHITECTURE_POSIX
+    /* Extract the protocol, hostname and port from the url */
+    UA_String hostname = UA_STRING_NULL;
+    UA_String path = UA_STRING_NULL;
+    UA_UInt16 port = 4840; /* default */
+    UA_StatusCode res = UA_parseEndpointUrl(&server->config.serverUrls[0], &hostname, &port, &path);
+    if(res != UA_STATUSCODE_GOOD) {
+        /* If the port cannot be extracted, the default port is used to generate the defaultPublisherId */
+        UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                     "PubSub generate defaultPublisherId failed. Could not extract the port from the url");
+    }
+    UA_UInt64 id = 0;
+    struct ifreq s;
+    strcpy(s.ifr_name, "eth0");
```

Comment:
We are not going to use platform-specific ioctl here.
It makes porting to embedded targets difficult.
In the future this can be retrieved from the eventloop.
For the time, comment this out and use random as default.

Reply:
That makes sense! An alternative would be to move the system call to the eventloop_posix_tcp and store the MAC address from there in the Connection Manager, for example.
Would that work?